File: crates/yavashark_test262/viewer/scripting/runtime.go
package scripting

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"sync"
	"time"
	"viewer/conf"
)

// Runtime manages the Deno subprocess for executing TypeScript scripts
type Runtime struct {
	cmd      *exec.Cmd
	stdin    io.WriteCloser
	stdout   *bufio.Reader
	stderr   io.ReadCloser
	mu       sync.Mutex // Protects running state and serializes Execute calls
	writeMu  sync.Mutex // Protects stdin writes (allows cancel during execute)
	running  bool
	scriptID int64
}

// ScriptResult represents the result of executing a script
type ScriptResult struct {
	Success bool            `json:"success"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   string          `json:"error,omitempty"`
}

// NewRuntime creates a new Deno runtime instance
func NewRuntime() *Runtime {
	return &Runtime{}
}

// Start launches the Deno subprocess
func (r *Runtime) Start() error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.running {
		return nil
	}

	// Find the runtime script path
	runtimeScript, err := r.findRuntimeScript()
	if err != nil {
		return fmt.Errorf("failed to find runtime script: %w", err)
	}

	// Build Deno command
	r.cmd = exec.Command(conf.DenoPath, "run",
		"--allow-net=localhost",
		"--allow-read",
		"--no-prompt",
		runtimeScript,
	)

	// Set up pipes
	r.stdin, err = r.cmd.StdinPipe()
	if err != nil {
		return fmt.Errorf("failed to create stdin pipe: %w", err)
	}

	stdout, err := r.cmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("failed to create stdout pipe: %w", err)
	}
	r.stdout = bufio.NewReader(stdout)

	r.stderr, err = r.cmd.StderrPipe()
	if err != nil {
		return fmt.Errorf("failed to create stderr pipe: %w", err)
	}

	// Start the process
	if err := r.cmd.Start(); err != nil {
		return fmt.Errorf("failed to start Deno: %w", err)
	}

	r.running = true

	// Start stderr reader goroutine
	go r.readStderr()

	return nil
}

// Stop terminates the Deno subprocess
func (r *Runtime) Stop() error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if !r.running {
		return nil
	}

	r.running = false

	// Close stdin to signal shutdown
	if r.stdin != nil {
		r.stdin.Close()
	}

	// Give process time to exit gracefully
	done := make(chan error, 1)
	go func() {
		done <- r.cmd.Wait()
	}()

	select {
	case <-done:
		// Process exited
	case <-time.After(2 * time.Second):
		// Force kill
		r.cmd.Process.Kill()
	}

	return nil
}

// sendRequest sends a JSON-RPC request to the Deno runtime (caller must hold writeMu)
func (r *Runtime) sendRequest(request *RPCRequest) error {
	reqBytes, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	if _, err := r.stdin.Write(append(reqBytes, '\n')); err != nil {
		return fmt.Errorf("failed to send request: %w", err)
	}

	return nil
}

// sendCancel sends a cancel request for the given execution ID
func (r *Runtime) sendCancel(executionID int64) error {
	r.writeMu.Lock()
	defer r.writeMu.Unlock()

	// Use a unique ID for the cancel request (negative to avoid collision)
	cancelRequest := NewRPCRequest(-executionID, "cancel", map[string]int64{"id": executionID})
	return r.sendRequest(cancelRequest)
}

// Execute runs a TypeScript script and returns the result
func (r *Runtime) Execute(ctx context.Context, script string, sessionID string, serverURL string) (*ScriptResult, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	if !r.running {
		return nil, fmt.Errorf("runtime not started")
	}

	// Increment script ID for tracking
	r.scriptID++
	id := r.scriptID

	// Create and send request
	request := NewRPCRequest(id, "execute", ExecuteParams{
		Script:    script,
		SessionID: sessionID,
		Timeout:   min(conf.ScriptTimeout, conf.MaxScriptTimeout),
		ServerURL: serverURL,
	})

	r.writeMu.Lock()
	err := r.sendRequest(request)
	r.writeMu.Unlock()
	if err != nil {
		return nil, err
	}

	// Read response with context cancellation support
	responseCh := make(chan []byte, 1)
	errCh := make(chan error, 1)

	go func() {
		// Read all responses until we get the one for our execute request
		// (we might get a cancel response first if cancellation races)
		for {
			line, err := r.stdout.ReadBytes('\n')
			if err != nil {
				errCh <- err
				return
			}

			// Check if this is the response for our execute request
			var response RPCResponse
			if err := json.Unmarshal(line, &response); err != nil {
				errCh <- fmt.Errorf("failed to unmarshal response: %w", err)
				return
			}

			// If this is the cancel response (negative ID), discard it
			if response.ID < 0 {
				continue
			}

			// If this matches our execute ID, return it
			if response.ID == id {
				responseCh <- line
				return
			}

			// Otherwise it's an unexpected response, keep reading
			// (shouldn't happen in normal operation)
		}
	}()

	// Set up cancellation goroutine
	cancelDone := make(chan struct{})
	defer close(cancelDone)

	go func() {
		select {
		case <-ctx.Done():
			// Context was cancelled, send cancel request to Deno
			if err := r.sendCancel(id); err != nil {
				fmt.Fprintf(os.Stderr, "[runtime] failed to send cancel: %v\n", err)
			}
		case <-cancelDone:
			// Execute completed, no need to cancel
		}
	}()

	select {
	case <-ctx.Done():
		// Wait a short time for the cancellation to propagate and get a response
		select {
		case line := <-responseCh:
			// Got a response (probably the cancel error response)
			var response RPCResponse
			if err := json.Unmarshal(line, &response); err != nil {
				return nil, fmt.Errorf("failed to unmarshal response: %w", err)
			}
			if response.Error != nil {
				return &ScriptResult{
					Success: false,
					Error:   response.Error.Message,
				}, nil
			}
			return &ScriptResult{
				Success: true,
				Result:  response.Result,
			}, nil
		case <-time.After(500 * time.Millisecond):
			// Cancellation timed out, return context error
			return nil, ctx.Err()
		case err := <-errCh:
			return nil, fmt.Errorf("failed to read response: %w", err)
		}
	case err := <-errCh:
		return nil, fmt.Errorf("failed to read response: %w", err)
	case line := <-responseCh:
		var response RPCResponse
		if err := json.Unmarshal(line, &response); err != nil {
			return nil, fmt.Errorf("failed to unmarshal response: %w", err)
		}

		if response.Error != nil {
			return &ScriptResult{
				Success: false,
				Error:   response.Error.Message,
			}, nil
		}

		return &ScriptResult{
			Success: true,
			Result:  response.Result,
		}, nil
	}
}

// findRuntimeScript locates the TypeScript runtime script
func (r *Runtime) findRuntimeScript() (string, error) {
	// Check relative to the viewer binary
	execPath, err := os.Executable()
	if err == nil {
		execDir := filepath.Dir(execPath)
		candidates := []string{
			filepath.Join(execDir, "scripting", "runtime.ts"),
			filepath.Join(execDir, "..", "scripting", "runtime.ts"),
		}
		for _, path := range candidates {
			if _, err := os.Stat(path); err == nil {
				return path, nil
			}
		}
	}

	// Check relative to working directory
	wd, err := os.Getwd()
	if err == nil {
		candidates := []string{
			filepath.Join(wd, "scripting", "runtime.ts"),
			filepath.Join(wd, "viewer", "scripting", "runtime.ts"),
		}
		for _, path := range candidates {
			if _, err := os.Stat(path); err == nil {
				return path, nil
			}
		}
	}

	return "", fmt.Errorf("runtime.ts not found")
}

// readStderr reads and logs stderr output from Deno
func (r *Runtime) readStderr() {
	scanner := bufio.NewScanner(r.stderr)
	for scanner.Scan() {
		fmt.Fprintf(os.Stderr, "[deno] %s\n", scanner.Text())
	}
}

// IsRunning returns whether the runtime is currently active
func (r *Runtime) IsRunning() bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.running
}

File: crates/yavashark_test262/viewer/scripting/runtime.ts
#!/usr/bin/env -S deno run --allow-net=localhost --allow-read

// YavaShark Test262 Scripting Runtime
// This is the main entry point for the Deno subprocess
// It reads JSON-RPC requests from stdin, executes scripts, and writes responses to stdout

import type { RPCRequest, RPCResponse, ExecuteParams } from "./types.ts";
import { ys, setServerUrl, setCurrentSession, resetOutput, getOutput, setAbortSignal } from "./api.ts";

// Track current execution for cancellation
let currentExecutionId: number | null = null;
let currentAbortController: AbortController | null = null;

// Mutex for writing responses (ensure responses aren't interleaved)
let writePromise: Promise<void> = Promise.resolve();

// Write a JSON-RPC response to stdout (serialized)
function writeResponse(response: RPCResponse): void {
  writePromise = writePromise.then(() => {
    const json = JSON.stringify(response);
    console.log(json);
  });
}

// Execute a user script using Function constructor
async function executeScript(params: ExecuteParams, abortSignal: AbortSignal): Promise<{ output: string[]; result: unknown }> {
  setCurrentSession(params.sessionId);
  resetOutput(); // Clear output from previous execution
  
  // Set server URL if provided
  if (params.serverUrl) {
    setServerUrl(params.serverUrl);
  }
  
  // Set abort signal for API calls
  setAbortSignal(abortSignal);
  
  // Wrap script in async function
  const asyncFn = new Function("ys", `
    return (async () => {
      ${params.script}
    })();
  `);

  // Create abort promise that rejects when cancelled
  const abortPromise = new Promise((_, reject) => {
    if (abortSignal.aborted) {
      reject(new DOMException("Script execution was cancelled", "AbortError"));
      return;
    }
    abortSignal.addEventListener("abort", () => {
      reject(new DOMException("Script execution was cancelled", "AbortError"));
    }, { once: true });
  });

  const result = await Promise.race([
    asyncFn(ys),
    abortPromise,
  ]);

  return {
    output: getOutput(),
    result,
  };
}

// Handle cancel request - returns response immediately, doesn't await
function handleCancelRequest(request: RPCRequest): RPCResponse {
  const params = request.params as { id?: number } | undefined;
  const targetId = params?.id;
  
  if (currentAbortController && (targetId === undefined || targetId === currentExecutionId)) {
    currentAbortController.abort();
    return {
      jsonrpc: "2.0",
      id: request.id,
      result: { cancelled: true, executionId: currentExecutionId },
    };
  } else {
    return {
      jsonrpc: "2.0",
      id: request.id,
      result: { cancelled: false, reason: "No matching execution to cancel" },
    };
  }
}

// Handle execute request - this is async and can be cancelled
async function handleExecuteRequest(request: RPCRequest): Promise<RPCResponse> {
  const params = request.params as ExecuteParams;
  
  // Set up cancellation
  currentExecutionId = request.id;
  currentAbortController = new AbortController();
  
  try {
    const { output, result } = await executeScript(params, currentAbortController.signal);
    
    // Build a combined result with output and return value
    let finalResult: unknown;
    
    if (output.length > 0 && result !== undefined) {
      finalResult = {
        output: output.join('\n'),
        result,
      };
    } else if (output.length > 0) {
      finalResult = {
        output: output.join('\n'),
      };
    } else if (result !== undefined) {
      if (typeof result === 'object' && result !== null) {
        finalResult = result;
      } else {
        finalResult = { value: result };
      }
    } else {
      finalResult = { message: "Script completed with no output" };
    }
    
    return {
      jsonrpc: "2.0",
      id: request.id,
      result: finalResult,
    };
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      return {
        jsonrpc: "2.0",
        id: request.id,
        error: {
          code: -32001,
          message: "Script execution was cancelled",
        },
      };
    } else {
      return {
        jsonrpc: "2.0",
        id: request.id,
        error: {
          code: -32000,
          message: error instanceof Error ? error.message : String(error),
        },
      };
    }
  } finally {
    currentExecutionId = null;
    currentAbortController = null;
    setAbortSignal(null);
  }
}

// Handle an incoming JSON-RPC request
function handleRequest(request: RPCRequest): Promise<RPCResponse> | RPCResponse {
  switch (request.method) {
    case "execute":
      return handleExecuteRequest(request);
    
    case "cancel":
      return handleCancelRequest(request);
    
    default:
      return {
        jsonrpc: "2.0",
        id: request.id,
        error: {
          code: -32601,
          message: `Method not found: ${request.method}`,
        },
      };
  }
}

// Main entry point - reads lines and dispatches requests concurrently
async function main(): Promise<void> {
  const decoder = new TextDecoder();
  const reader = Deno.stdin.readable.getReader();
  let buffer = "";
  
  // Track the current execute promise so we can ensure proper ordering
  let currentExecutePromise: Promise<void> | null = null;

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        if (!line.trim()) continue;
        
        let request: RPCRequest;
        try {
          request = JSON.parse(line) as RPCRequest;
        } catch (error) {
          writeResponse({
            jsonrpc: "2.0",
            id: 0,
            error: {
              code: -32700,
              message: `Parse error: ${error instanceof Error ? error.message : String(error)}`,
            },
          });
          continue;
        }

        // Handle cancel requests immediately (synchronously)
        if (request.method === "cancel") {
          const response = handleRequest(request);
          writeResponse(response as RPCResponse);
          continue;
        }

        // For execute requests, we need to wait for any previous execute to finish
        // to maintain ordering, but we dispatch them asynchronously so cancel can interrupt
        if (request.method === "execute") {
          // Wait for previous execute to complete (if any)
          if (currentExecutePromise) {
            await currentExecutePromise;
          }
          
          // Start new execute and track it
          currentExecutePromise = (async () => {
            const responsePromise = handleRequest(request);
            const response = responsePromise instanceof Promise 
              ? await responsePromise 
              : responsePromise;
            writeResponse(response);
          })();
          
          // Don't await here - let the main loop continue to read cancel requests
          // The response will be written when the execute completes
        } else {
          // Other requests are handled synchronously
          const response = handleRequest(request);
          if (response instanceof Promise) {
            writeResponse(await response);
          } else {
            writeResponse(response);
          }
        }
      }
    }

    // Handle remaining buffer
    if (buffer.trim()) {
      try {
        const request = JSON.parse(buffer) as RPCRequest;
        const response = handleRequest(request);
        if (response instanceof Promise) {
          writeResponse(await response);
        } else {
          writeResponse(response);
        }
      } catch (error) {
        writeResponse({
          jsonrpc: "2.0",
          id: 0,
          error: {
            code: -32700,
            message: `Parse error: ${error instanceof Error ? error.message : String(error)}`,
          },
        });
      }
    }
    
    // Wait for any pending execute to complete
    if (currentExecutePromise) {
      await currentExecutePromise;
    }
  } finally {
    reader.releaseLock();
  }
}

main().catch(console.error);
